datalad/distribution/publish.py

```diff
@@ -85,115 +86,230 @@ class Publish(Interface):
+                                         path=relpath(path, start=subds.path),
+                                         dest_url=dest_url,
+                                         dest_pushurl=dest_pushurl,
+                                         with_data=with_data,
+                                         recursive=recursive)
+
+        # now, we know, we have to operate on ds. So, ds needs to be installed,
+        # since we cannot publish anything from a not installed dataset, can we?
+        # (But may be just the existence of ds.repo is important here.)
+        if not ds.is_installed():
+            raise ValueError("No installed dataset found at {0}.".format(ds.path))
+        assert(ds.repo is not None)
+
+        # TODO: For now we can deal with a sibling(remote) name given by `dest`
+        # only. Figure out, when to allow for passing a local path or URL
+        # directly and what to do in that case.
+
+        # Note: we need an upstream remote, if there's none given. We could
+        # wait for git push to complain, but we need to explicitly figure it
+        # out for pushing annex branch anyway and we might as well fail right
+        # here.
+
+        # keep original dest in case it's None for passing to recursive calls:
+        dest_resolved = dest
+        if dest is None:
+            # check for tracking branch's remote:
             try:
-                # Note: tracking branch actually defined bei entry "merge"
-                # PLUS entry "remote"
                 std_out, std_err = \
-                    handle_repo._git_custom_command('',
-                                                    ["git", "config", "--get", "branch.{active_branch}.merge".format(active_branch=handle_repo.git_get_active_branch())])
+                    ds.repo._git_custom_command('',
+                                                ["git", "config", "--get", "branch.{active_branch}.remote".format(active_branch=ds.repo.git_get_active_branch())],
+                                                expect_fail=True)
             except CommandError as e:
                 if e.code == 1 and e.stdout == "":
-                    # no tracking branch:
-                    set_upstream = True
+                    std_out = None
                 else:
                     raise
+            if std_out:
+                dest_resolved = std_out.strip()
+            else:
+                # we have no remote given and no upstream => fail
+                raise RuntimeError("No known default target for "
+                                   "publication and none given.")
 
+        # upstream branch needed for update (merge) and subsequent push,
+        # in case there is no.
+        set_upstream = False
+        try:
+            # Note: tracking branch actually defined bei entry "merge"
+            # PLUS entry "remote"
+            std_out, std_err = \
+                ds.repo._git_custom_command('',
```

Comment:
FWIW, might be worth to use convention in such cases to move all arguments under indented just by one level, i.e.
```python
ds.repo._git_custom_command(
    '',
    ["git....
```
since now can't even see all the code without horizontal scrolling